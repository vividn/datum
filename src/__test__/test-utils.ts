import { CouchDbError } from "../errors";
import {
  DataOnlyDocument,
  DatumData,
  DatumDocument,
  DatumMetadata,
  EitherDocument,
  EitherPayload,
} from "../documentControl/DatumDocument";
import Mock = jest.Mock;
import { DateTime, Settings } from "luxon";
import { parseTimeStr } from "../time/parseTimeStr";
import { now } from "../time/timeUtils";
import { connectDb } from "../auth/connectDb";
import * as connectDbModule from "../auth/connectDb";
import { defaultIdComponents } from "../ids/defaultIdComponents";
import { buildIdStructure } from "../ids/buildIdStructure";
import { defaults } from "../input/defaults";
import { assembleId } from "../ids/assembleId";
import * as newHumanIdModule from "../meta/newHumanId";

// eslint-disable-next-line @typescript-eslint/no-empty-function
export const pass = (): void => {};
export const fail = (): never => {
  throw Error;
};

export const mockDocMissingError: CouchDbError = {
  status: 404,
  name: "not_found",
  error: "not_found",
  reason: "missing",
  message: "missing",
};

export const mockDocDeletedError: CouchDbError = {
  status: 404,
  message: "deleted",
  name: "not_found",
  error: "not_found",
  reason: "deleted",
};

export const mockMissingNamedViewError: CouchDbError = {
  status: 404,
  name: "not_found",
  error: "not_found",
  reason: "missing_named_view",
  message: "missing_named_view",
};

export async function resetTestDb(
  db: PouchDB.Database & {
    __opts?: PouchDB.Configuration.DatabaseConfiguration;
    _destroyed?: boolean;
    _closed?: boolean;
  },
): Promise<PouchDB.Database> {
  await db.destroy().catch(pass);
  // nasty hack to reopen closed database
  delete db._destroyed;
  delete db._closed;
  return db.constructor(db.name, db.__opts);
}

export function testDbLifecycle(
  dbName: string,
): PouchDB.Database<EitherPayload> {
  const db = connectDb({ db: dbName });

  beforeEach(async () => {
    await resetTestDb(db);
    jest.spyOn(connectDbModule, "connectDb").mockReturnValue(db);
  });

  afterEach(async () => {
    await db.destroy().catch(pass);
  });

  return db;
}

export function mockedLogLifecycle(): Mock {
  const originalLog = console.log;
  const mockedLog = jest.fn() as Mock;

  beforeEach(async () => {
    console.log = mockedLog;
  });

  afterEach(async () => {
    console.log = originalLog;
    mockedLog.mockReset();
  });

  return mockedLog;
}

const originalNowFn = Settings.now;
const nowStack: (() => number)[] = [];
export function setNow(timeStr: string): DateTime {
  const parsedTime = parseTimeStr({ timeStr, referenceTime: DateTime.now() });
  const newNow = () => parsedTime.toMillis();
  Settings.now = newNow;
  return parsedTime;
}

export function pushNow(timeStr: string): DateTime {
  nowStack.push(Settings.now);
  return setNow(timeStr);
}

export function popNow(): DateTime {
  const lastNow = nowStack.pop();
  if (lastNow === undefined) {
    throw new Error("tried to pop non existent now");
  }
  Settings.now = lastNow;
  return now();
}

export function restoreNow(): DateTime {
  Settings.now = originalNowFn;
  nowStack.length = 0;
  return now();
}

export function at<A extends any[], O>(
  timeStr: string,
  fn: (...args: A) => Promise<O>,
): (...args: A) => Promise<O> {
  return async (...args: A): Promise<O> => {
    pushNow(timeStr);
    const returnVal = await fn(...args);
    popNow();
    return returnVal;
  };
}

// TODO: Transition all tests to use makeDoc where appropriate
export function makeDoc(
  data: DatumData,
  meta: DatumMetadata | false = {},
  include_rev = false,
): EitherDocument {
  let doc: EitherDocument;
  if (meta === false) {
    doc = { ...data } as DataOnlyDocument;
  } else {
    doc = { ...data, meta } as DatumDocument;
  }

  meta = meta || {};
  if (doc._id === undefined) {
    let idStructure: string;
    if (meta?.idStructure) {
      idStructure = meta.idStructure;
    } else {
      const { defaultIdParts, defaultPartitionParts } = defaultIdComponents({
        data,
      });
      idStructure = buildIdStructure({
        idParts: defaultIdParts,
        delimiter: defaults.idDelimiter,
        partition: defaultPartitionParts,
      });
    }
    doc._id = assembleId({ payload: doc, idStructure });
  }

  if (include_rev) {
    doc._rev = "1-foo";
  }
  return doc;
}

export async function deterministicHumanIds(seed?: number): Promise<void> {
  let a = seed || 20231018;
  function random() {
    const x = Math.sin(a++) * 10000;
    return x - Math.floor(x);
  }
  function mockNewHumanId(): string {
    return random().toString(36).slice(2) + random().toString(36).slice(2);
  }

  beforeEach(() => {
    a = seed || 20231018;
    jest
      .spyOn(newHumanIdModule, "newHumanId")
      .mockImplementation(mockNewHumanId);
  });
}

// export async function generateSampleDay(dateStr = "2022-08-14") {
//
//   setNow(dateStr);
//   at('8:30', addCmd)({field: "sleep", })
//
//   popNow();
// }
