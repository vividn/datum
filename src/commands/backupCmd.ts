import { MainDatumArgs } from "../input/mainArgs";
import { connectDb } from "../auth/connectDb";
import * as fs from "fs";
import { DateTime } from "luxon";
import { dbArgs } from "../input/dbArgs";
import { ArgumentParser } from "argparse";
import { parseIfNeeded } from "../utils/parseIfNeeded";

export const backupArgs = new ArgumentParser({
  add_help: false,
});
// TODO: make this optional and then use stdout if not specified
backupArgs.add_argument("filename", {
  help: "file to write the backup to",
  type: "str",
});
backupArgs.add_argument("--overwrite", {
  help: "overwrite file if it exists",
  action: "store_true",
});

export const backupCmdArgs = new ArgumentParser({
  description: "Backup db, outputs to stdout",
  prog: "datum backup",
  usage: "%(prog)s <filename>",
  parents: [backupArgs, dbArgs],
});

export type BackupCmdArgs = MainDatumArgs & {
  filename: string;
  overwrite?: boolean;
};

export async function backupCmd(
  args: BackupCmdArgs | string | string[],
  preparsed?: Partial<BackupCmdArgs>,
): Promise<void> {
  args = parseIfNeeded(backupCmdArgs, args, preparsed);
  if (!args.overwrite && fs.existsSync(args.filename)) {
    throw new Error("File exists, overwrite with --overwrite");
  }
  const db = connectDb(args);
  const allDocs = (await db.allDocs({ include_docs: true })).rows.map(
    ({ doc }) => doc,
  );
  // TODO: Also backup and restore attachments (even though using attachments is considered not best practice)
  const backupTime = DateTime.utc().toISO() as string;
  const backupData = JSON.stringify({ backupTime, docs: allDocs }, null, 0);
  fs.writeFileSync(args.filename, backupData);
}
