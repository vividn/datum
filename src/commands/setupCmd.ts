import { Argv } from "yargs";
import { BaseDatumArgs } from "../input/baseYargs";
import { setupDatumViews } from "../views/setupDatumViews";
import { connectDb } from "../auth/connectDb";
import { Show } from "../output/output";

export const command = "setup";
export const desc = "setup the database for use with datum";

export type SetupCmdArgs = BaseDatumArgs & {
  projectDir?: string;
};

export function builder(yargs: Argv): Argv {
  return yargs.options({
    "project-dir": {
      describe:
        "where to look for additional DatumViews to setup in the database",
      type: "string",
    },
  });
}

export async function setupCmd(args: SetupCmdArgs): Promise<void> {
  const db = connectDb(args);
  const show: Show = args.showAll ? Show.All : args.show ?? Show.None;
  await setupDatumViews({ db, show: show, projectDir: args.projectDir });
}
