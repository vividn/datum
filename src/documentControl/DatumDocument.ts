import {
  isIsoDateOrTime,
  isoDate,
  isoDatetime,
  isoDuration,
} from "../time/timeUtils";
import { WithRequired } from "../utils/utilityTypes";

export type GenericData<T = unknown> = T & {
  [key: string]: any;
};

export type DatumData<T = unknown> = GenericData<T> & {
  occurTime?: isoDatetime | isoDate;
  occurUtcOffset?: number;
  dur?: "start" | "end" | isoDuration;
};

export type OccurredData<T = unknown> = WithRequired<
  DatumData<T>,
  "occurTime" | "occurUtcOffset"
>;

export function isOccurredData(
  data: DatumData | OccurredData
): data is OccurredData {
  const occurTime = data.occurTime;
  return occurTime !== undefined && isIsoDateOrTime(occurTime);
}

export type DatumMetadata = {
  utcOffset?: number;
  createTime?: isoDatetime;
  modifyTime?: isoDatetime; //TODO: turn into an array of times
  idStructure?: string;
  random?: number;
  humanId?: string;
  // [key: string]: any;
};

export type DatumPayload<T = unknown> = {
  _id?: string;
  _rev?: string;
  data: DatumData<T>;
  meta: DatumMetadata;
};

export type DatumDocument<T = unknown> = DatumPayload<T> & {
  _id: string;
  _rev: string;
};

export type DataOnlyPayload<T = unknown> = {
  _id?: string;
  _rev?: string;
} & DatumData<T>;

export type DataOnlyDocument<T = unknown> = DataOnlyPayload<T> & {
  _id: string;
  _rev: string;
};

export type EitherPayload<T = unknown> = DatumPayload<T> | DataOnlyPayload<T>;
export type EitherDocument<T = unknown> =
  | DatumDocument<T>
  | DataOnlyDocument<T>;

export function isDatumDocument(
  doc: DatumDocument | DataOnlyDocument
): doc is DatumDocument {
  return (
    (doc as DatumDocument).data !== undefined &&
    (doc as DatumDocument).meta !== undefined
  );
}

export function isDatumPayload(
  payload: DatumPayload | DataOnlyPayload
): payload is DatumPayload {
  return (
    (payload as DatumPayload).data !== undefined &&
    (payload as DatumPayload).meta !== undefined
  );
}
