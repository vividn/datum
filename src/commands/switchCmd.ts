import { occurCmd, OccurCmdArgs } from "./occurCmd";
import { EitherDocument } from "../documentControl/DatumDocument";
import { flexiblePositional } from "../input/flexiblePositional";
import { durationArgs, DurationArgs } from "../input/durationArgs";

import { DatumState, normalizeState } from "../state/normalizeState";
import { ArgumentParser } from "argparse";
import { parseIfNeeded } from "../utils/parseIfNeeded";
import { dataArgs } from "../input/dataArgs";
import { timeArgs } from "../input/timeArgs";
import { fieldArgs } from "../input/fieldArgs";
import { dbArgs } from "../input/dbArgs";
import { outputArgs } from "../input/outputArgs";
import { newDocArgs } from "./addCmd";
import { inferType } from "../utils/inferType";

export const stateArgs = new ArgumentParser({
  add_help: false,
});
stateArgs.add_argument("state", {
  help: "the state to switch to, it defaults to true--equivalent to start",
  nargs: "?",
  default: "true",
});
stateArgs.add_argument("--last-state", "-L", {
  help: "manually specify the last state being transitioned out of",
  dest: "lastState",
});

export const switchArgs = new ArgumentParser({
  add_help: false,
  parents: [fieldArgs, stateArgs, durationArgs, newDocArgs, timeArgs, dataArgs],
});

export const switchCmdArgs = new ArgumentParser({
  description: "switch states of a given field",
  prog: "datum switch",
  usage: `%(prog)s <field> <state> [<duration OR .> [data..]]
  %(prog)s --moment <field> <state> [data..]
  %(prog)s <field> -k <reqKey> -k <optKey>=[defaultValue] ... <state> <duration OR .> <reqValue> [optValue] [data..]
`,
  parents: [switchArgs, dbArgs, outputArgs],
});

export type StateArgs = {
  state?: DatumState | string;
  lastState?: DatumState | string;
};
export type SwitchCmdArgs = OccurCmdArgs & DurationArgs & StateArgs;

export async function switchCmd(
  args: SwitchCmdArgs | string | string[],
  preparsed?: Partial<SwitchCmdArgs>,
): Promise<EitherDocument> {
  args = parseIfNeeded(switchCmdArgs, args, preparsed);
  flexiblePositional(
    args,
    "duration",
    "dur=",
    args.moment || args.omitTimestamp,
  );
  flexiblePositional(args, "state", "state=true");
  args.cmdData ??= {};
  if (args.moment) {
    args.cmdData.dur = null;
  }
  if (args.lastState) {
    const lastState =
      typeof args.lastState === "string"
        ? inferType(args.lastState)
        : args.lastState;
    if (lastState !== undefined) {
      args.cmdData.lastState = normalizeState(lastState);
    }
  }

  return await occurCmd(args);
}
