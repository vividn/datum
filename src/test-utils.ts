import Nano, { DocumentScope } from "nano";
import { CouchDbError } from "./errors";
import { EitherPayload } from "./documentControl/DatumDocument";
import * as connectDb from "./auth/connectDb";
import Mock = jest.Mock;
import { DateTime, Settings } from "luxon";
import { parseTimeStr } from "./time/parseTimeStr";
import { now } from "./time/timeUtils";

// eslint-disable-next-line @typescript-eslint/no-empty-function
export const pass = (): void => {};
export const fail = (): never => {
  throw Error;
};

export const testNano = Nano(`http://admin:password@localhost:5983`);

export const mockDocMissingError: CouchDbError = {
  scope: "couch",
  statusCode: 404,
  errid: "non_200",
  description: "missing",
  error: "not_found",
  reason: "missing",
};

export const mockDocDeletedError: CouchDbError = {
  scope: "couch",
  statusCode: 404,
  errid: "non_200",
  description: "deleted",
  error: "not_found",
  reason: "deleted",
};

export const mockMissingNamedViewError: CouchDbError = {
  scope: "couch",
  statusCode: 404,
  errid: "non_200",
  description: "missing_named_view",
  error: "not_found",
  reason: "missing_named_view",
};

export async function resetTestDb(
  dbName: string
): Promise<DocumentScope<EitherPayload>> {
  const maxTries = 3;
  let tries = 0;
  await testNano.db.destroy(dbName).catch(pass);
  while (tries++ <= maxTries) {
    await testNano.db.destroy(dbName).catch(pass);
    await testNano.db.create(dbName).catch(pass);
    if ((await testNano.db.list()).includes(dbName)) {
      return testNano.use(dbName) as DocumentScope<EitherPayload>;
    }
  }
  throw Error(`Unable to reset database after ${maxTries} attempts`);
}

export function testDbLifecycle(dbName: string): DocumentScope<EitherPayload> {
  const db = testNano.use(dbName) as DocumentScope<EitherPayload>;

  beforeEach(async () => {
    await resetTestDb(dbName);
    jest.spyOn(connectDb, "connectDb").mockImplementation(async () => db);
  });

  afterEach(async () => {
    await testNano.db.destroy(dbName).catch(pass);
  });

  return db;
}

export function mockedLogLifecycle(): Mock {
  const originalLog = console.log;
  const mockedLog = jest.fn() as Mock;

  beforeEach(async () => {
    console.log = mockedLog;
  });

  afterEach(async () => {
    console.log = originalLog;
    mockedLog.mockReset();
  });

  return mockedLog;
}

const originalNowFn = Settings.now;
const nowStack: (() => number)[] = [];
export function setNow(timeStr: string): DateTime {
  const parsedTime = parseTimeStr({ timeStr });
  const newNow = () => parsedTime.toMillis();
  Settings.now = newNow;
  return parsedTime;
}

export function pushNow(timeStr: string): DateTime {
  nowStack.push(Settings.now);
  return setNow(timeStr);
}

export function popNow(): DateTime {
  const lastNow = nowStack.pop();
  if (lastNow === undefined) {
    throw new Error("tried to pop non existent now");
  }
  Settings.now = lastNow;
  return now();
}

export function restoreNow(): DateTime {
  Settings.now = originalNowFn;
  nowStack.length = 0;
  return now();
}

export function at<A extends any[], O>(
  timeStr: string,
  fn: (...args: A) => O
): (...args: A) => O {
  return (...args: A): O => {
    pushNow(timeStr);
    const returnVal: O = fn(...args);
    popNow();
    return returnVal;
  };
}

// export async function generateSampleDay(dateStr = "2022-08-14") {
//
//   setNow(dateStr);
//   at('8:30', addCmd)({field: "sleep", })
//
//   popNow();
// }
