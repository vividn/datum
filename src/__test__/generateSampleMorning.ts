import { endCmd } from "../commands/endCmd";
import { occurCmd } from "../commands/occurCmd";
import { startCmd } from "../commands/startCmd";
import { switchCmd } from "../commands/switchCmd";
import { popNow, pushNow, setNow } from "./test-utils";

export async function generateSampleMorning(date: string): Promise<void> {
  pushNow(`8:30 ${date}`);
  await endCmd("sleep");
  setNow("+5");
  await startCmd("sleep -y -t 2320");
  setNow("8:40");
  await switchCmd("project german");
  await switchCmd("text fiction_book");
  setNow("9:10");
  await endCmd("project");
  await endCmd("text");
  setNow("9:30");
  await switchCmd("environment outside");
  await startCmd("stretch");
  setNow("+7");
  await endCmd("stretch");
  await startCmd("run");
  setNow("+30");
  await endCmd("run distance=5.4");
  await startCmd("stretch");
  setNow("+8");
  await endCmd("stretch");
  await switchCmd("environment home");
  setNow("+3");
  await occurCmd("pushup amount=10");
  setNow("10:30");
  await switchCmd("environment outside 5");
  setNow("11");
  await occurCmd("caffeine amount=100 -c coffee");
  popNow();
}
