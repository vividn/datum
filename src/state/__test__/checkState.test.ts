import { restoreNow, setNow, testDbLifecycle } from "../../__test__/test-utils";
import { setupCmd } from "../../commands/setupCmd";
import { switchCmd } from "../../commands/switchCmd";
import { DatumDocument } from "../../documentControl/DatumDocument";
import {
  checkState,
  LastStateError,
  OverlappingBlockError,
  RepeatedStateError,
  StateErrorSummary,
} from "../checkState";

const noErrors: StateErrorSummary = { ok: true, errors: [] };
describe("checkState", () => {
  const db = testDbLifecycle("check_state_test");

  beforeEach(() => {
    setupCmd({});
    setNow("2024-09-05 10:15:00");
  });

  afterEach(() => {
    restoreNow();
  });

  it("returns no errors for a field with no data", async () => {
    const retVal = await checkState({ db, field: "test" });
    expect(retVal).toEqual(noErrors);
  });

  it("returns no errors if every state transistions properly to the next state", async () => {
    await switchCmd("field state1");
    setNow("+5");
    const doc2 = (await switchCmd("field state2")) as DatumDocument; // automatically adds the lastState
    expect(doc2.data.lastState).toEqual("state1");
    setNow("+10");
    const doc3 = (await switchCmd(
      "field state3 --last-state state2",
    )) as DatumDocument; // explicit lastState
    expect(doc3.data.lastState).toEqual("state2");

    const doc4 = (await switchCmd(
      "field tempState dur=4 -t -4",
    )) as DatumDocument; // puts a block in the middle of state2
    expect(doc4.data.lastState).toEqual("state2");

    const retVal = await checkState({ db, field: "field" });
    expect(retVal).toEqual(noErrors);
  });

  it("throws a LastStateError if lastState does not reflect the last state correctly", async () => {
    await switchCmd("field state1");
    setNow("+5");
    await switchCmd("field newState --last-state wrongState");

    await expect(checkState({ db, field: "field" })).rejects.toThrow(
      LastStateError,
    );
    const errors = await checkState({ db, field: "field", failOnError: false });
    expect(errors.ok).toBe(false);
    expect(errors.errors).toHaveLength(1);
    expect(errors.errors[0]).toBeInstanceOf(LastStateError);
    expect(errors.errors[0]).toMatchSnapshot();
  });

  it("throws a RepeatedStateError if the same state is repeated", async () => {
    await switchCmd("field state1");
    setNow("+5");
    await switchCmd("field state1");

    const errors = await checkState({ db, field: "field", failOnError: false });
    expect(errors.ok).toBe(false);
    expect(errors.errors).toHaveLength(1);
    expect(errors.errors[0].name).toEqual("RepeatedStateError");
    expect(errors.errors[0]).toMatchSnapshot();
  });

  it("throws an OverlappingBlockError if a state change block is inserted and overlaps another state change", async () => {
    setNow("10:45");
    await switchCmd("project emails");
    setNow("11");
    await switchCmd("project meetings");
    setNow("11:10");
    await switchCmd("project overlapping dur=15");

    await expect(checkState({ db, field: "project" })).rejects.toThrow(
      OverlappingBlockError,
    );
    const errors = await checkState({
      db,
      field: "project",
      failOnError: false,
    });
    expect(errors.ok).toBe(false);
    expect(errors.errors).toHaveLength(1);
    expect(errors.errors[0]).toBeInstanceOf(OverlappingBlockError);
    expect(errors.errors).toMatchSnapshot();
  });

  it("throws an OverlappingBlockError if a state change is added in the middle of an existing block", async () => {
    setNow("11");
    await switchCmd("project emails");
    await switchCmd("project frontend dur=15 -t 11:30");
    await switchCmd("project backend -t 11:20"); // starts in the middle of the frontend block

    await expect(checkState({ db, field: "project" })).rejects.toThrow(
      OverlappingBlockError,
    );
    const errors = await checkState({
      db,
      field: "project",
      failOnError: false,
    });
    expect(errors.ok).toBe(false);
    expect(errors.errors).toHaveLength(1);
    expect(errors.errors[0]).toBeInstanceOf(OverlappingBlockError);
    expect(errors.errors[0]).toMatchSnapshot();
  });

  it("throws an OverlappingBlockError if two blocks overlap an edge", async () => {
    setNow("12");
    await switchCmd("project emails");
    await switchCmd("project frontend dur=20 -t 12:30");
    await switchCmd("project backend dur=15 -t 12:40");

    await expect(checkState({ db, field: "project" })).rejects.toThrow(
      OverlappingBlockError,
    );
    const errors = await checkState({
      db,
      field: "project",
      failOnError: false,
    });
    expect(errors.ok).toBe(false);
    expect(errors.errors).toHaveLength(2);
    expect(errors.errors[0]).toBeInstanceOf(OverlappingBlockError);
    expect(errors.errors[1]).toBeInstanceOf(OverlappingBlockError);
    expect(errors.errors[0]).not.toEqual(errors.errors[1]);
    expect(errors.errors).toMatchSnapshot();
  });

  it("throws an OverlappingBlockError if one block is nested in another, and has the wrong lastState", async () => {
    // checkState only notices fully nested blocks as being a problem when the lastState is incorrect. However, overlapping blocks in any form is unsupported and would be caught by checkOverlappingBlocks
    setNow("12");
    await switchCmd("project emails");
    setNow("12:30");
    await switchCmd("project frontend dur=20");
    await switchCmd("project backend -t -5 dur=10 --last-state notFrontend");

    await expect(checkState({ db, field: "project" })).rejects.toThrow(
      OverlappingBlockError,
    );
    const errors = await checkState({
      db,
      field: "project",
      failOnError: false,
    });
    expect(errors.ok).toBe(false);
    expect(errors.errors).toHaveLength(2);
    expect(errors.errors[0]).toBeInstanceOf(OverlappingBlockError);
    expect(errors.errors[1]).toBeInstanceOf(OverlappingBlockError);
    expect(errors.errors[0]).not.toEqual(errors.errors[1]);
    expect(errors.errors).toMatchSnapshot();
  });

  it("detects a LastStateError that occurs immediately after an OverlappingBlockError", async () => {
    setNow("2024-09-11 16:30");
    await switchCmd("machine on");
    await switchCmd("machine reboot -t 16:50 10");
    await switchCmd("machine ready -t 16:45");
    await switchCmd(
      "machine off -t 16:55 --last-state not-ready-but-should-be",
    );

    const errors = await checkState({
      db,
      field: "machine",
      failOnError: false,
    });
    expect(errors.ok).toBe(false);
    expect(errors.errors).toHaveLength(2);
    expect(errors.errors[0]).toBeInstanceOf(OverlappingBlockError);
    expect(errors.errors[1]).toBeInstanceOf(LastStateError);
  });

  it("correctly handles based off of a given startTime and/or endTime", async () => {
    setNow("2024-09-05T13:00:00Z");
    await switchCmd("field state1");
    setNow("+5");
    await switchCmd("field state2");
    setNow("+10");
    await switchCmd("field state3 --last-state wrongState");
    setNow("2024-09-05T14:00:00Z");
    await switchCmd("field state4");
    setNow("+5");
    await switchCmd("field overlappingBlock dur=10");
    setNow("2024-09-05T15:00:00Z");
    await switchCmd("field repeatedState");
    setNow("+10");
    await switchCmd("field repeatedState");
    setNow("2024-09-05T16:00:00Z");
    await switchCmd("field state5");
    setNow("+10");
    await switchCmd("field state6");

    await expect(
      checkState({ db, field: "field", startTime: "2024-09-05T13:00:00Z" }),
    ).rejects.toThrow(LastStateError);
    await expect(
      checkState({ db, field: "field", startTime: "2024-09-05T13:55:00Z" }),
    ).rejects.toThrow(OverlappingBlockError);
    await expect(
      checkState({ db, field: "field", startTime: "2024-09-05T14:45:00Z" }),
    ).rejects.toThrow(RepeatedStateError);
    await expect(
      checkState({
        db,
        field: "field",
        startTime: "2024-09-05T12:00:00Z",
        endTime: "2024-09-05T13:10:00Z",
      }),
    ).resolves.toEqual(noErrors);
    await expect(
      checkState({
        db,
        field: "field",
        endTime: "2024-09-05T13:10:00Z",
      }),
    ).resolves.toEqual(noErrors);
    await expect(
      checkState({
        db,
        field: "field",
        startTime: "2024-09-05T15:55:00Z",
      }),
    ).resolves.toEqual(noErrors);

    const allErrors = await checkState({
      db,
      field: "field",
      failOnError: false,
    });
    expect(allErrors.ok).toBe(false);
    expect(allErrors.errors).toHaveLength(3);
    expect(allErrors.errors).toMatchSnapshot();
  });

  it("throws a LastStateError if the lastState of the first ever row for a field is not null", async () => {
    await switchCmd("field newState --last-state false");
    await expect(checkState({ db, field: "field" })).rejects.toThrow(
      LastStateError,
    );

    await switchCmd("field2 newState --last-state null");
    await expect(checkState({ db, field: "field2" })).resolves.toEqual(
      noErrors,
    );
  });

  it("detects if the first row does not have the correct lastState when a startTime is given", async () => {
    setNow("2024-09-05T13:00:00Z");
    await switchCmd("field newState --last-state false");
    await expect(
      checkState({ db, field: "field", startTime: "2024-09-05T12:55:00Z" }),
    ).rejects.toThrow(LastStateError);

    setNow("2024-09-05T13:30:00Z");
    await expect(
      checkState({ db, field: "field", startTime: "2024-09-05T13:25:00Z" }),
    ).resolves.toEqual(noErrors);
    await switchCmd("field newState --last-state incorrect");
    await expect(
      checkState({ db, field: "field", startTime: "2024-09-05T13:25:00Z" }),
    ).rejects.toThrow(LastStateError);
  });
});

describe("checkState --fix", () => {
  const db = testDbLifecycle("check_state_fix_test");

  beforeEach(() => {
    setupCmd({});
    setNow("2024-09-10 15:00:00");
  });

  afterEach(() => {
    restoreNow();
  });

  it("fixes a LastStateError by updating the lastState", async () => {
    await switchCmd("field state1");
    setNow("+5");
    await switchCmd("field state2");
    setNow("+10");
    const doc3 = await switchCmd("field state3 --last-state wrongState");

    const retVal = await checkState({ db, field: "field", fix: true });
    expect(retVal).toEqual(noErrors);

    const newDoc3 = (await db.get(doc3._id)) as DatumDocument;
    expect(newDoc3.data.lastState).toEqual("state2");

    const newCheck = await checkState({ db, field: "field" });
    expect(newCheck).toEqual(noErrors);
  });

  it("can fix multiple LastStateErrors", async () => {
    setNow("2024-09-16 17:45");
    const doc1 = await switchCmd("field state1");
    setNow("+5");
    const doc2 = await switchCmd("field state2 -L wrongState1");
    setNow("+5");
    const doc3 = await switchCmd("field state3 --last-state wrongState2");

    const retVal = await checkState({
      db,
      field: "field",
      fix: true,
    });
    expect(retVal).toEqual(noErrors);

    const retVal2 = await checkState({ db, field: "field" });
    expect(retVal2).toEqual(noErrors);

    const newDoc1 = (await db.get(doc1._id)) as DatumDocument;
    const newDoc2 = (await db.get(doc2._id)) as DatumDocument;
    const newDoc3 = (await db.get(doc3._id)) as DatumDocument;

    expect(newDoc1.data.lastState).toEqual(null);
    expect(newDoc1).toEqual(doc1);
    expect(newDoc2.data.lastState).toEqual("state1");
    expect(newDoc3.data.lastState).toEqual("state2");
  });

  it("can fix many LastStateErros caused by duration blocks built upon a false state", async () => {
    setNow("2024-09-16 17:30");
    const doc1 = await switchCmd("field falseState");
    setNow("18:00");
    const doc2 = await switchCmd("field tempState1 dur=10");
    setNow("18:15");
    const doc3 = await switchCmd("field tempState2 dur=5");
    setNow("18:30");
    const doc4 = await switchCmd("field tempState3 dur=8");
    const doc5 = await switchCmd("field realBaseState -t 17:45");

    const retVal = await checkState({ db, field: "field", fix: true });
    expect(retVal).toEqual(noErrors);

    const newDoc1 = (await db.get(doc1._id)) as DatumDocument;
    const newDoc2 = (await db.get(doc2._id)) as DatumDocument;
    const newDoc3 = (await db.get(doc3._id)) as DatumDocument;
    const newDoc4 = (await db.get(doc4._id)) as DatumDocument;
    const newDoc5 = (await db.get(doc5._id)) as DatumDocument;

    expect(newDoc1).toEqual(doc1);
    expect(newDoc5).toEqual(doc5);

    expect(newDoc2.data.lastState).toEqual("realBaseState");
    expect(newDoc3.data.lastState).toEqual("realBaseState");
    expect(newDoc4.data.lastState).toEqual("realBaseState");

    const retVal2 = await checkState({ db, field: "field" });
    expect(retVal2).toEqual(noErrors);
  });
});
