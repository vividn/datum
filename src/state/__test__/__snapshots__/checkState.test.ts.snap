// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`checkState correctly handles based off of a given startTime and/or endTime 1`] = `
[
  [LastStateError: field 2024-09-05T13:15:00.000Z: lastState does not match actual last state. ids: [field:2024-09-05T13:05:00.000Z, field:2024-09-05T13:15:00.000Z]],
  [OverlappingBlockError: field 2024-09-05T14:00:00.000Z: State changes within a block. ids: [field:2024-09-05T14:05:00.000Z, field:2024-09-05T14:00:00.000Z ]}],
  [RepeatedStateError: field 2024-09-05T15:10:00.000Z: state "repeatedState" is repeated. ids: [field:2024-09-05T15:00:00.000Z, field:2024-09-05T15:10:00.000Z]],
]
`;

exports[`checkState throws a LastStateError if lastState does not reflect the last state correctly 1`] = `[LastStateError: field 2024-09-05T10:20:00.000Z: lastState does not match actual last state. ids: [field:2024-09-05T10:15:00.000Z, field:2024-09-05T10:20:00.000Z]]`;

exports[`checkState throws a RepeatedStateError if the same state is repeated 1`] = `[RepeatedStateError: field 2024-09-05T10:20:00.000Z: state "state1" is repeated. ids: [field:2024-09-05T10:15:00.000Z, field:2024-09-05T10:20:00.000Z]]`;

exports[`checkState throws an OverlappingBlockError if a state change block is inserted and overlaps another state change 1`] = `
[
  [OverlappingBlockError: project 2024-09-05T11:00:00.000Z: State changes within a block. ids: [project:2024-09-05T11:10:00.000Z, project:2024-09-05T11:00:00.000Z ]}],
]
`;

exports[`checkState throws an OverlappingBlockError if a state change is added in the middle of an existing block 1`] = `[OverlappingBlockError: project 2024-09-05T11:20:00.000Z: State changes within a block. ids: [project:2024-09-05T11:30:00.000Z, project:2024-09-05T11:20:00.000Z ]}]`;

exports[`checkState throws an OverlappingBlockError if one block is nested in another, and has the wrong lastState 1`] = `
[
  [OverlappingBlockError: project 2024-09-05T12:15:00.000Z: Block starts within a block. ids: [project:2024-09-05T12:30:00.000Z, project:2024-09-05T12:25:00.000Z ]}],
  [OverlappingBlockError: project 2024-09-05T12:30:00.000Z: Block ends within a block. ids: [project:2024-09-05T12:25:00.000Z, project:2024-09-05T12:30:00.000Z ]}],
]
`;

exports[`checkState throws an OverlappingBlockError if two blocks overlap an edge 1`] = `
[
  [OverlappingBlockError: project 2024-09-05T12:25:00.000Z: Block starts within a block. ids: [project:2024-09-05T12:30:00.000Z, project:2024-09-05T12:40:00.000Z ]}],
  [OverlappingBlockError: project 2024-09-05T12:40:00.000Z: Block ends within a block. ids: [project:2024-09-05T12:30:00.000Z, project:2024-09-05T12:40:00.000Z ]}],
]
`;
